File: app/src/main/java/de/cineaste/android/activity/AbstractSearchActivity.java
package de.cineaste.android.activity;

import android.app.ActivityOptions;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.util.Pair;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.ProgressBar;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.lang.reflect.Type;

import de.cineaste.android.R;
import de.cineaste.android.listener.ItemClickListener;
import de.cineaste.android.network.NetworkCallback;
import de.cineaste.android.network.NetworkResponse;
import de.cineaste.android.util.DateAwareGson;


public abstract class AbstractSearchActivity extends AppCompatActivity implements ItemClickListener {

    final Gson gson = new DateAwareGson().getGson();
    RecyclerView recyclerView;
    ProgressBar progressBar;
    private SearchView searchView;
    private String searchText;

    protected abstract Intent getIntentForDetailActivity(long itemId);
    protected abstract int getLayout();
    protected abstract void initAdapter();
    protected abstract RecyclerView.Adapter getListAdapter();
    protected abstract void getSuggestions();
    protected abstract void searchRequest( String searchQuery);
    protected abstract Type getListType();
    protected abstract Runnable getRunnable(final String json, final Type listType);


    @Override
    public void onItemClickListener(long itemId, View[] views) {
        Intent intent = getIntentForDetailActivity(itemId);

        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            ActivityOptions options = ActivityOptions.makeSceneTransitionAnimation(this,
                    Pair.create(views[0], "card"),
                    Pair.create(views[1], "poster"));
            this.startActivity(intent, options.toBundle());
        } else {
            this.startActivity(intent);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getLayout());

        initToolbar();

        if (savedInstanceState != null) {
            searchText = savedInstanceState.getString("query", null).replace("+", " ");
        }

        progressBar = findViewById(R.id.progressBar);
        recyclerView = findViewById(R.id.search_recycler_view);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this);
        initAdapter();
        recyclerView.setItemAnimator(new DefaultItemAnimator());

        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(getListAdapter());

        getSuggestions();
    }

    private void initToolbar() {
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        if (searchView != null) {
            if (!TextUtils.isEmpty(searchText)) {
                outState.putString("query", searchText);
            }
        }
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onPause() {
        super.onPause();
        Bundle outState = new Bundle();
        outState.putString("query", searchText);
        onSaveInstanceState(outState);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {

        MenuInflater menuInflater = getMenuInflater();
        menuInflater.inflate(R.menu.search_menu, menu);

        MenuItem searchItem = menu.findItem(R.id.action_search);

        if (searchItem != null) {
            searchView = (SearchView) searchItem.getActionView();
            searchView.setFocusable(true);
            searchView.setIconified(false);
            searchView.requestFocusFromTouch();
            searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
                @Override
                public boolean onQueryTextSubmit(String query) {
                    return false;
                }

                @Override
                public boolean onQueryTextChange(String query) {
                    if (!query.isEmpty()) {
                        query = query.replace(" ", "+");
                        progressBar.setVisibility(View.VISIBLE);

                        scheduleSearchRequest(query);

                        searchText = query;
                    } else {
                        getSuggestions();
                    }
                    return false;
                }

            });
            if (!TextUtils.isEmpty(searchText))
                searchView.setQuery(searchText, false);
        }

        return super.onCreateOptionsMenu(menu);
    }

    private void scheduleSearchRequest(String query) {
        searchView.removeCallbacks(getSearchRunnable(query));
        searchView.postDelayed(getSearchRunnable(query), 500);
    }

    @NonNull
    private Runnable getSearchRunnable(final String searchQuery) {
        return new Runnable() {
            @Override
            public void run() {
                searchRequest(searchQuery);
            }
        };
    }

    private void showNetworkError() {
        Snackbar snackbar = Snackbar
                .make(recyclerView, R.string.noInternet, Snackbar.LENGTH_LONG);
        snackbar.show();
    }


    @Override
    protected void onStop() {
        super.onStop();
        View view = getCurrentFocus();
        if (view != null) {
            InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
            if (imm == null) {
                return;
            }
            imm.hideSoftInputFromInputMethod(view.getWindowToken(), 0);
        }

    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return super.onSupportNavigateUp();
    }

    @NonNull
    NetworkCallback getNetworkCallback() {
        return new NetworkCallback() {
            @Override
            public void onFailure() {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        showNetworkError();
                    }
                });
            }

            @Override
            public void onSuccess(NetworkResponse response) {
                JsonParser parser = new JsonParser();
                JsonObject responseObject =
                        parser.parse(response.getResponseReader()).getAsJsonObject();
                final String json = responseObject.get("results").toString();
                final Type listType = getListType();

                runOnUiThread(getRunnable(json, listType));
            }
        };
    }
}

File: app/src/main/kotlin/de/cineaste/android/activity/AbstractSearchActivity.kt
package de.cineaste.android.activity

import android.app.ActivityOptions
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.text.TextUtils
import android.util.Pair
import android.view.Menu
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.ProgressBar
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import com.google.gson.JsonParser
import de.cineaste.android.R
import de.cineaste.android.listener.ItemClickListener
import de.cineaste.android.network.NetworkCallback
import de.cineaste.android.network.NetworkResponse
import de.cineaste.android.util.DateAwareGson
import kotlinx.coroutines.Dispatchers.Main
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.lang.reflect.Type

abstract class AbstractSearchActivity : AppCompatActivity(), ItemClickListener {

    internal val gson = DateAwareGson.gson
    internal lateinit var recyclerView: RecyclerView
    internal lateinit var progressBar: ProgressBar
    private lateinit var searchView: SearchView
    private var searchText: String? = null
    protected abstract val layout: Int
    protected abstract val listAdapter: RecyclerView.Adapter<*>
    protected abstract val listType: Type

    val networkCallback: NetworkCallback
        get() = object : NetworkCallback {
            override fun onFailure() {
                GlobalScope.launch(Main) { showNetworkError() }
            }

            override fun onSuccess(response: NetworkResponse) {
                val responseObject = JsonParser.parseReader(response.responseReader).asJsonObject
                val json = responseObject.get("results").toString()
                val listType = listType

                GlobalScope.launch(Main) { getRunnable(json, listType).run() }
            }
        }

    protected abstract fun getIntentForDetailActivity(itemId: Long): Intent
    protected abstract fun initAdapter()
    protected abstract fun getSuggestions()
    protected abstract fun searchRequest(searchQuery: String)
    protected abstract fun getRunnable(json: String, listType: Type): Runnable

    override fun onItemClickListener(itemId: Long, views: Array<View>) {
        val intent = getIntentForDetailActivity(itemId)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            val options = ActivityOptions.makeSceneTransitionAnimation(
                this,
                Pair.create(views[0], "card"),
                Pair.create(views[1], "poster")
            )
            this.startActivity(intent, options.toBundle())
        } else {
            this.startActivity(intent)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layout)

        initToolbar()

        if (savedInstanceState != null) {
            searchText = savedInstanceState.getString("query", "").replace("+", " ")
        }

        progressBar = findViewById(R.id.progressBar)
        recyclerView = findViewById(R.id.search_recycler_view)
        val layoutManager = LinearLayoutManager(this)
        val divider = ContextCompat.getDrawable(recyclerView.context, R.drawable.divider)
        val itemDecor = DividerItemDecoration(
            recyclerView.context,
            layoutManager.orientation
        )
        divider?.let {
            itemDecor.setDrawable(it)
        }
        recyclerView.addItemDecoration(itemDecor)
        initAdapter()
        recyclerView.itemAnimator = DefaultItemAnimator()

        recyclerView.layoutManager = layoutManager
        recyclerView.adapter = listAdapter

        progressBar.visibility = View.VISIBLE
        getSuggestions()
    }

    private fun initToolbar() {
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        val actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)
    }

    public override fun onSaveInstanceState(outState: Bundle) {
        if (!TextUtils.isEmpty(searchText)) {
            outState.putString("query", searchText)
        }
        super.onSaveInstanceState(outState)
    }

    public override fun onPause() {
        super.onPause()
        val outState = Bundle()
        outState.putString("query", searchText)
        onSaveInstanceState(outState)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {

        val menuInflater = menuInflater
        menuInflater.inflate(R.menu.search_menu, menu)

        val searchItem = menu.findItem(R.id.action_search)

        if (searchItem != null) {
            searchView = searchItem.actionView as SearchView
            searchView.isFocusable = true
            searchView.isIconified = false
            searchView.requestFocusFromTouch()
            searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(query: String): Boolean {
                    return false
                }

                override fun onQueryTextChange(query: String): Boolean {
                    var myQuery = query
                    if (myQuery.isNotEmpty()) {
                        myQuery = myQuery.replace(" ", "+")
                        progressBar.visibility = View.VISIBLE

                        scheduleSearchRequest(myQuery)

                        searchText = myQuery
                    } else {
                        getSuggestions()
                    }
                    return false
                }
            })
            if (!TextUtils.isEmpty(searchText))
                searchView.setQuery(searchText, false)
        }

        return super.onCreateOptionsMenu(menu)
    }

    private fun scheduleSearchRequest(query: String) {
        searchView.removeCallbacks(getSearchRunnable(query))
        searchView.postDelayed(getSearchRunnable(query), 500)
    }

    private fun getSearchRunnable(searchQuery: String): Runnable {
        return Runnable { searchRequest(searchQuery) }
    }

    private fun showNetworkError() {
        val snackBar = Snackbar
            .make(recyclerView, R.string.noInternet, Snackbar.LENGTH_LONG)
        snackBar.show()
    }

    override fun onStop() {
        super.onStop()
        val view = currentFocus
        if (view != null) {
            val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromInputMethod(view.windowToken, 0)
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return super.onSupportNavigateUp()
    }
}
